Improve the UI/UX of Promgen

hoangpn opened the pull request:
Rendering too much data on some templates may cause high load on the web browser as well as the server side. We make some small changes to improve the UI/UX of these screens:

### 1. Add pagination to Search page, RulesList page and HostList page classes to prevent rendering too much data:

- Search page:
   - 20 elements per content per page (_For example, if we have 30 services and 20 projects, the result when searching all will have 20 services and 20 projects on **the first page**; and **the second page** only shows next 10 services_)
   - total_page = MAX(total_page of each content)

- RulesList page:
   - 50 elements per page

- HostList page:
   - 50 elements per page

### 2. Replace _pagination.html_ with _pagination_short.html_ for a more compact display and easier page navigation:
AS-IS:
<img width="1175" alt="image" src="https://github.com/user-attachments/assets/496e38a1-17d9-436a-a060-ca574afa4b20" />

TO-BE:
![image](https://github.com/user-attachments/assets/ffed6f48-d2b0-4870-a051-1cd66c54618d)

### 3. Prevent searching with empty keywords:
<img width="395" alt="image" src="https://github.com/user-attachments/assets/78630d05-32f0-4fd9-8c7c-76af6efaf444" />

### 4. Fix some minor display-related bugs.

vincent-olivert-riera commented:
@hoangpn, very nice improvement, thank you!

Now, I have one first suggestion to do. Let's look at the list of changes you have implemented:

>- Add pagination to some ListView classes to prevent rendering too much data.
>- Replace pagination.html with pagination_short.html for a more compact display
and easier page navigation.
>- Add an expandable area for the host_list of a farm when there are more than 10 hosts.
>- Prevent searching with empty keywords.
>- Fix some minor display-related bugs.

Each one of those changes are logically independent, therefore they should be in separate commits. Please do that (you will need to `push --force`, no problem), and then I will start reviewing your changes commit by commit.

Again, thank you very much :bow:

vincent-olivert-riera commented on promgen/views.py:
```diff
@@ -1218,7 +1230,20 @@ def get(self, request):
+                # Since we run each query separately, there are many paginator objects.
+                # However, we only want to display a single navigation. Therefore, we want to use
+                # the largest paginator object to render the page navigation.
+                if ("page_obj" not in context
+                        or context[
+                            "page_obj"].paginator.num_pages < page_target.paginator.num_pages):
```
Please, format this in a nicer way:

```
if (
    "page_obj" not in context
    or context["page_obj"].paginator.num_pages < page_target.paginator.num_pages
):
```

vincent-olivert-riera reviewed (changes requested):
Looks good to me, but could you please address the comment about the `if` formatting?

We can leave the messages translation for another PR if you want.

hoangpn commented:
@vincent-olivert-riera
Thank you for reviewing my PR. I fixed the formatting problem. I also added the translation message.
Please review the fixup commits. Feel free to ask me to change if necessary! 🙇

vincent-olivert-riera commented:
@hoangpn , looks good. Please rebase on master while squashing the fixup commits:

```
git rebase -i master --autosquash
```

Then `push --force` for final approval.

hoangpn commented:
Sorry, I accidentlly click the re-request button. 🙇

vincent-olivert-riera commented:
@hoangpn , during my last review I found that commit 799d898d19e2da1701d6c40724410d1565a9812c does not look good enough to merge it.
The most important thing in this PR is the fix for the empty searches, which has been done correctly. So, in order to not delay the merging of this PR, I suggest dropping commit 799d898d19e2da1701d6c40724410d1565a9812c from it.

That feature can be added in a different PR, but it will need to be implemented in a different way. We can discuss about that later, but basically, those kind of things are better done using CSS and Javascript. And there must be a "show more" button which becomes "show less" after expanding the list. And given that kind of expandable/collapsible list functionality can be reused, the best way to implement it would be by creating a new Vue component.

As you can see, the changes are not trivial, therefore my suggestion is to simply drop commit 799d898d19e2da1701d6c40724410d1565a9812c so we can merge this PR and move on.

vincent-olivert-riera reviewed (approved):
Looks good to me.
